### DISCUSSION TITLE
Some Queries regarding Fluvio

### DISCUSSION THREAD

**Discussion by cloudcompute** [fluvio-community/fluvio#2980] [Q&A] [answered]
Hi

a. Does it support publish-subscribe and/or request-reply pattern(s)? Pub-Sub is 1 producer and multiple consumers reading at the same time? Is it push or pull?

b. Does it offer support for Transactions like Kafka does i.e. writing different events to multiple topics in a single transaction?

b. Can we use Fluvio for Event Sourcing? What is the (approx) maximum data (in GB) that the underlying data retention store can hold without any impact on performance?

c. Fluvio has source and sink connectors for Kafka, what role is Kafka playing here? Is Fluvio not a pub-sub messaging system, therefore you are offering connectivity to Kafka?

d. Looks like Smart Modules are somewhat equivalent to Apache Flink? Hopefully it also address use cases like analysing clickstreams, fraud and anomaly detection,

e. Is it stable enough so that we can use it in production?

f. What is the difference between Infinyon Cloud and Fluvio Open Source? Is there any comparison matrix?

Thanks

**Comment by sehz** [comment, marked as answer]
Hi @cloudcompute,  good questions, responses here:

a. Fluvio like Kafa supports `publish-subscribe` pattern.  Not sure `request-reply`,  maybe you can give better details.   Fluvio is fully async so it's not like push-pull which is more for sync API.

b. Fluvio doesn't support transaction yet.

b. Yes Fluvio is ideal for Event Sourcing.   Size of data should not impact performance.

c. Fluvio's connector can connect to many different sources and sinks.  Kafka is just another source or sink.  This is ideal for if you have existing Kafka installation and wants to source data from it.   Fluvio can fully perform entire data processing pipeline without any other existern dependencies.

d.  SmartModule is next generation of data transformation engine built on WebAssembly technology that goes beyond what Flink is capable.  It should support clickstreams, fault and anomaly detection use cases but we haven't fully tested yet.  We love have you try it out.

e. Yes it is stable and is running in production workload for customers. 

f.  Fluvio Open source is perfect for running dev workloads.  As open source, there is no SLA other than thru Github project.  InfinyOnCloud is fully managed offering with support.  In addition it has dashboard to to monitor usages.

**Comment by cloudcompute** [comment]
Hi @sehz 

Thank you for the quick and detailed response.

[This link](https://docs.nats.io/nats-concepts/overview/compare-nats) compares the popular messaging platforms.

Taken from the same site that describes the request-reply paradigm.
> Request-Reply is a common pattern in modern distributed systems. A request is sent, and the application either waits on the response with a certain timeout, or receives a response asynchronously. 

My honest `feedback regarding Fluvio`.. The most powerful thing it offers is the reduction in deployment complexity and operational costs. As of now, it lacks some features to make it as powerful as Kafka and few others... i am saying this based on the comparison described in the link shared above.

With Regards
